File: src/main/java/fr/tikione/c2e/service/web/AbstractReader.java
package fr.tikione.c2e.service.web;

import fr.tikione.c2e.model.web.Auth;
import org.jsoup.Jsoup;
import org.jsoup.nodes.Document;
import org.jsoup.nodes.Element;
import org.jsoup.safety.Whitelist;
import org.jsoup.select.Elements;

import java.io.IOException;
import java.util.Arrays;

public abstract class AbstractReader {
    
    public final String CPC_BASE_URL = "https://www.canardpc.com";
    public final String CPC_LOGIN_FORM_POST_URL = CPC_BASE_URL + "/user/login";
    public final String CPC_MAG_NUMBER_BASE_URL = CPC_BASE_URL + "/numero/_NUM_";
    public static final String CUSTOMTAG_EM_START = "__c2e_emStart__";
    public static final String CUSTOMTAG_EM_END = "__c2e_emEnd__";
    public static final String CUSTOMTAG_STRONG_START = "__c2e_strongStart__";
    public static final String CUSTOMTAG_STRONG_END = "__c2e_strongEnd__";
    private static final String EM_START = "<em>";
    private static final String EM_END = "</em>";
    private static final String STRONG_START = "<span class=\"title\">";
    private static final String STRONG_END = "</span>";
    
    /** UserAgent to include when scrapping CanardPC website. This is not mandatory, but a good practice. */
    public static final String UA = "fr.tikione.c2e";
    
    /** Replace unbreakable spaces by regular spaces, and apply trim. */
    public String clean(String str) {
        while (str.contains(String.valueOf((char) 160))) {
            str = str.replace(String.valueOf((char) 160), " ");
        }
        String res = Jsoup.clean(str.trim(), Whitelist.none());
        if (res.toUpperCase().endsWith("TWITTER FACEBOOK EMAIL")) {
            res = res.substring(0, res.toUpperCase().lastIndexOf("TWITTER FACEBOOK EMAIL")); // remove social buttons
        }
        return res.trim();
    }
    
    /** Get a remote document. */
    public Document queryUrl(Auth auth, String url) throws IOException {
        return Jsoup.connect(url)
                .cookies(auth.getCookies())
                .userAgent(UA)
                .get();
    }
    
    public String text(Element elt) {
        return elt == null ? null : clean((elt.text()));
    }
    
    /** Get text and keep some HTML styling tags: {@code em} and {@code strong}. */
    public String richText(Element elt) {
        if (elt == null) {
            return null;
        }
        String text = elt.html();
        int idx = 0;
        while (text.indexOf(EM_START, idx) >= 0) {
            idx = text.indexOf(EM_START, idx) + EM_START.length();
            int idxEnd = text.indexOf(EM_END, idx) + EM_END.length();
            if (idx < idxEnd) {
                text = text.substring(0, idx) + CUSTOMTAG_EM_START + text.substring(idx, idxEnd) + CUSTOMTAG_EM_END + text.substring(idxEnd);
            }
        }
        idx = 0;
        while (text.indexOf(STRONG_START, idx) >= 0) {
            idx = text.indexOf(STRONG_START, idx) + STRONG_START.length();
            int idxEnd = text.indexOf(STRONG_END, idx) + STRONG_END.length();
            if (idx < idxEnd) {
                text = text.substring(0, idx) + CUSTOMTAG_STRONG_START + text.substring(idx, idxEnd) + CUSTOMTAG_STRONG_END + text.substring(idxEnd);
            }
        }
        return clean(text);
    }
    
    public String text(Elements... elt) {
        if (elt == null || elt.length == 0) {
            return null;
        }
        StringBuilder buff = new StringBuilder();
        Arrays.stream(elt).forEach(elements -> buff.append(elements == null ? "" : clean(elements.text())));
        return buff.toString();
    }
    
    public String attr(Element elt, String attr) {
        return elt == null ? null : elt.attr(attr);
    }
    
    public String attr(Elements elt, String attr) {
        return elt == null ? null : elt.attr(attr);
    }
    
    public String attr(String baseUrl, Elements elt, String attr) {
        return elt == null || elt.attr(attr) == null || elt.attr(attr).trim().isEmpty() ? null : baseUrl + elt.attr(attr);
    }
}

File: src/main/kotlin/fr/tikione/c2e/core/service/web/AbstractReader.kt
package fr.tikione.c2e.core.service.web

import fr.tikione.c2e.core.model.web.Auth
import org.jsoup.Jsoup
import org.jsoup.nodes.Document
import org.jsoup.nodes.Element
import org.jsoup.safety.Whitelist
import org.jsoup.select.Elements
import java.io.IOException

abstract class AbstractReader {

    /** Replace unbreakable spaces by regular spaces, and apply trim.  */
    fun clean(strToClean: String): String {
        var str = strToClean
        while (str.contains(160.toChar().toString())) {
            str = str.replace(160.toChar().toString(), " ")
        }
        var res = Jsoup.clean(str.trim { it <= ' ' }, Whitelist.none())
        if (res.toUpperCase().endsWith("TWITTER FACEBOOK EMAIL")) {
            res = res.substring(0, res.toUpperCase().lastIndexOf("TWITTER FACEBOOK EMAIL")) // remove social buttons
        }
        return res.trim { it <= ' ' }
    }

    /** Get a remote document.  */
    @Throws(IOException::class)
    fun queryUrl(auth: Auth, url: String): Document = Jsoup.connect(url)
            .cookies(auth.cookies)
            .userAgent(UA)
            .execute().parse()

    fun text(elt: Element?): String? = if (elt == null) null else clean(elt.text())

    /** Get text and keep some HTML styling tags: `em` and `strong`.  */
    fun richText(elt: Element?): String? {
        if (elt == null) {
            return null
        }
        var text = elt.html()
        var idx = 0
        while (text.indexOf(EM_START, idx) >= 0) {
            idx = text.indexOf(EM_START, idx) + EM_START.length
            val idxEnd = text.indexOf(EM_END, idx) + EM_END.length
            if (idx < idxEnd) {
                text = text.substring(0, idx) + CUSTOMTAG_EM_START + text.substring(idx, idxEnd) + CUSTOMTAG_EM_END + text.substring(idxEnd)
            }
        }
        idx = 0
        while (text.indexOf(STRONG_START, idx) >= 0) {
            idx = text.indexOf(STRONG_START, idx) + STRONG_START.length
            val idxEnd = text.indexOf(STRONG_END, idx) + STRONG_END.length
            if (idx < idxEnd) {
                text = text.substring(0, idx) + CUSTOMTAG_STRONG_START + text.substring(idx, idxEnd) + CUSTOMTAG_STRONG_END + text.substring(idxEnd)
            }
        }
        return clean(text)
    }

    fun text(vararg elt: Elements): String? {
        if (elt.isEmpty()) {
            return null
        }
        val buff = StringBuilder()
        elt.iterator().forEach { elements -> buff.append(clean(elements.text())) }
        return buff.toString()
    }

    fun attr(elt: Element?, attr: String): String? = elt?.attr(attr)

    fun attr(elt: Elements?, attr: String): String? = elt?.attr(attr)

    fun attr(baseUrl: String, elt: Elements?, attr: String): String? =
            if (elt?.attr(attr) == null || elt.attr(attr).trim { it <= ' ' }.isEmpty()) null else baseUrl + elt.attr(attr)

    companion object {
        const val CPC_BASE_URL = "https://www.canardpc.com"
        const val CPC_LOGIN_FORM_POST_URL = "$CPC_BASE_URL/user/login"
        const val CPC_MAG_NUMBER_BASE_URL = "$CPC_BASE_URL/numero/_NUM_"
        const val CUSTOMTAG_EM_START = "__c2e_emStart__"
        const val CUSTOMTAG_EM_END = "__c2e_emEnd__"
        const val CUSTOMTAG_STRONG_START = "__c2e_strongStart__"
        const val CUSTOMTAG_STRONG_END = "__c2e_strongEnd__"
        private const val EM_START = "<em>"
        private const val EM_END = "</em>"
        private const val STRONG_START = "<span class=\"title\">"
        private const val STRONG_END = "</span>"

        /** UserAgent to include when scrapping CanardPC website. This is not mandatory, but a good practice.  */
        const val UA = "fr.tikione.c2e"
    }
}
